dartagnan/src/main/java/com/dat3m/dartagnan/wmm/relation/base/RelRMW.java

```diff
@@ -136,58 +101,65 @@ public TupleSet getMaxTupleSet(){
-        
-        // Encode base (not exclusive pairs) RMW
-        TupleSet origEncodeTupleSet = encodeTupleSet;
-        encodeTupleSet = new TupleSet(Sets.intersection(encodeTupleSet, baseMaxTupleSet));
-        BooleanFormula enc = super.encodeApprox(ctx);
-        encodeTupleSet = origEncodeTupleSet;
-
-        // Encode RMW for exclusive pairs
-		BooleanFormula unpredictable = bmgr.makeFalse();
-        for(Thread thread : task.getProgram().getThreads()) {
-            for (Event store : thread.getCache().getEvents(storeExclFilter)) {
-            	BooleanFormula storeExec = bmgr.makeFalse();
-                for (Event load : thread.getCache().getEvents(loadExclFilter)) {
-                    if (load.getCId() < store.getCId()) {
-
-                        // Encode if load and store form an exclusive pair
-                    	BooleanFormula isPair = exclPair(load, store, ctx);
-                    	BooleanFormula isExecPair = bmgr.and(isPair, store.exec());
-                        enc = bmgr.and(enc, bmgr.equivalence(isPair, pairingCond(thread, load, store, ctx)));
-
-                        // If load and store have the same address
-                        BooleanFormula sameAddress = generalEqual(((MemEvent)load).getMemAddressExpr(), ((MemEvent)store).getMemAddressExpr(), ctx);
-                        unpredictable = bmgr.or(unpredictable, bmgr.and(isExecPair, bmgr.not(sameAddress)));
-
-                        // Relation between exclusive load and store
-                        enc = bmgr.and(enc, bmgr.equivalence(this.getSMTVar(load, store, ctx), bmgr.and(isExecPair, sameAddress)));
-
-                        // For ARMv8, the store can be executed if addresses mismatch, but behaviour is "constrained unpredictable"
-                        // The implementation does not include all possible unpredictable cases: in case of address
-                        // mismatch, addresses of read and write are unknown, i.e. read and write can use any address.
-                        // For RISCV and Power, addresses should match.
-                        BooleanFormula succCond = store.is(Tag.MATCHADDRESS) ? bmgr.and(isPair, sameAddress) : isPair;
-                        storeExec = bmgr.or(storeExec, succCond);
-                    }
+        BooleanFormula enc = bmgr.makeTrue();
+        BooleanFormula unpredictable = bmgr.makeFalse();
+        for(Event store : task.getProgram().getCache().getEvents(
+                FilterIntersection.get(FilterBasic.get(Tag.WRITE), FilterBasic.get(Tag.EXCL)))) {
+            checkState(store instanceof MemEvent, "non-memory event participating in '" + getName() + "'");
+            BooleanFormula storeExec = bmgr.makeFalse();
+            for(Tuple t : maxTupleSet.getBySecond(store)) {
+                MemEvent load = (MemEvent) t.getFirst();
+                if(minTupleSet.contains(t)) {
+                    storeExec = bmgr.or(storeExec, load.exec());
+                    continue;
```

Comment:
The pairing conditions still need to hold in this case. In particular, under RISC-V and Power the store must fail if there is an address mismatch (that is why there is the MATCHADDRESS Tag which is added to RISC-V and Power SC events, but not to ARMv8 one's).

Reply:
This means that `minTupleSet` is currently greater than it should be, even for ARMv8, whose encoding still entails equal addresses.  Seems like an unprecedented case, is there no test on ARMv8 with reserve/exclusive instructions with different addresses that requires a consistent execution with even the store executed?
Shouldn't the entire `rmw` be a subset of `loc`?

Reply:
I don't think we have litmus tests for ARMv8 that have LL/SC pairs accessing different addresses. We do have such tests for RISC-V though (and some of those break in this PR).
I think in all cases, a rmw-edge should only be established when the addresses match. In the case of Power/Risc-V this is justified because the store has to fail if addresses mismatch. In the case of ARMv8, mismatching addresses give rise to a kind of undefined behavior which we do not fully model (who the hell does that write mismatching LL/SC pairs to begin with?!).

So I would say a LL/SC pair is in the `minTupleSet` iff
- there is never an intervening operation
- In the case of ARMv8, both events are must-alias.

Note that the address condition is irrelevant in the case of RISC-V/Power, because if the addresses mismatch, the store has to fail (i.e. not get executed). So in every execution that contains both events, they have to go to the same address and hence form a rmw-pair.
But care has to be taken that this failure condition gets encoded under all circumstances, because it is part of the anarchic semantics. It's a little unfortunate, that `RelRMW` encodes both the rmw-edges as seen from the memory model as well as the anarchic semantics of LL/SC operations.

Reply:
The issue of allowing execution of the SC under POWER/RISC-V even if its address mismatches with the LL is still there.
In other words, an SC that is not part of any rmw-edge should not get executed at all (under POWER and RISC-V).

Reply:
You were right.  As long as `RelRMW` encodes the anarchic semantics of store-exclusives, that conditional is obsolete.